app/models/reports/repository.rb

```diff
@@ -1,64 +1,143 @@
+    end
+
+    # Returns the full range of assigned patient counts for a Region. We do this via one SQL query for each Region, because its
+    # fast and easy via the underlying query.
+    smart_memoize def full_assigned_patients_counts
+      items = regions.map { |region| RegionEntry.new(region, :cumulative_assigned_patients_count, with_exclusions: with_exclusions) }
+      cache.fetch_multi(*items, force: force_cache?) { |entry|
+        AssignedPatientsQuery.new.count(entry.region, :month, with_exclusions: with_exclusions)
+      }
+    end
+
+    # Return the running total of cumulative assigned patient counts.
+    smart_memoize def cumulative_assigned_patients_count
+      full_assigned_patients_counts.each_with_object({}) do |(region_entry, patient_counts), totals|
+        region_slug = region_entry.region.slug
+        totals[region_slug] = Hash.new(0)
+        next if patient_counts.empty?
+        full_range = (patient_counts.keys.first..periods.end)
+        full_range.each do |period|
+          previous_total = totals[region_slug][period.previous]
+          current_amount = patient_counts[period] || 0
+          totals[region_slug][period] += previous_total + current_amount
+        end
+      end
+    end
+
+    smart_memoize def controlled_patients_count
+      cached_query(__method__) do |entry|
+        control_rate_query.controlled(entry.region, entry.period, with_exclusions: with_exclusions).count
       end
     end
 
-    def uncontrolled_patients_count
-      cached_query(:uncontrolled_patients_count) do |item|
-        control_rate_query.uncontrolled(item.region, item.period, with_exclusions: with_exclusions).count
+    smart_memoize def uncontrolled_patients_count
+      cached_query(__method__) do |entry|
+        control_rate_query.uncontrolled(entry.region, entry.period, with_exclusions: with_exclusions).count
       end
     end
 
-    def no_bp_measure_count
-      cached_query(:no_bp_measure_count) do |item|
-        no_bp_measure_query.call(item.region, item.period, with_exclusions: with_exclusions)
+    smart_memoize def controlled_patient_rates
+      cached_query(__method__) do |entry|
+        controlled = controlled_patients_count[entry.region.slug][entry.period]
+        total = cumulative_assigned_patients_count[entry.region.slug][entry.period]
+        percentage(controlled, total)
       end
     end
 
-    # Generate all necessary cache keys for a calculation, then yield to the block with every Item.
+    smart_memoize def uncontrolled_patient_rates
+      cached_query(__method__) do |entry|
+        controlled = uncontrolled_patients_count[entry.region.slug][entry.period]
+        total = cumulative_assigned_patients_count[entry.region.slug][entry.period]
+        percentage(controlled, total)
+      end
+    end
+
+    smart_memoize def missed_visits_count
+      cached_query(__method__) do |entry|
+        no_bp_measure_query.call(entry.region, entry.period, with_exclusions: with_exclusions)
```

Comment:
To calculate missed visits we need to do
```
missed_visits = registrations - visited_without_bp_taken - controlled - uncontrolled
```

Not sure if the right thing is being called here. We should name the result of `no_bp_measure_query` as `visited_without_bp_taken` in any case.

Reply:
Yup, you're right.  I've fixed this up to have the correct name now.